phpBB/phpbb/storage/file_tracker.php

```diff
@@ -0,0 +1,198 @@
+<?php
+/**
+ *
+ * This file is part of the phpBB Forum Software package.
+ *
+ * @copyright (c) phpBB Limited <https://www.phpbb.com>
+ * @license GNU General Public License, version 2 (GPL-2.0)
+ *
+ * For full copyright and license information, please see
+ * the docs/CREDITS.txt file.
+ *
+ */
+
+namespace phpbb\storage;
+
+use phpbb\cache\driver\driver_interface as cache;
+use phpbb\db\driver\driver_interface as db;
+use phpbb\storage\exception\storage_exception;
+
+class file_tracker
+{
+
+	/**
+	 * @var db
+	 */
+	protected $db;
+
+	/**
+	 * Cache driver
+	 * @var cache
+	 */
+	protected $cache;
+
+	/**
+	 * @var string
+	 */
+	protected $storage_table;
+
+	/**
+	 * Constructor
+	 *
+	 * @param db								$db
+	 * @param cache								$cache
+	 * @param string							$storage_table
+	 */
+	public function __construct(db $db, cache $cache, string $storage_table)
+	{
+		$this->db = $db;
+		$this->cache = $cache;
+		$this->storage_table = $storage_table;
+	}
+
+	/**
+	 * Track file in database
+	 *
+	 * @param string $storage	Storage name
+	 * @param string $path		The target file
+	 * @param int $size			Size in bytes
+	 */
+	public function track_file(string $storage, string $path, int $size): void
+	{
+		$sql_ary = array(
+			'file_path'		=> $path,
+			'storage'		=> $storage,
+			'filesize'		=> $size,
+		);
+
+		$sql = 'INSERT INTO ' . $this->storage_table . $this->db->sql_build_array('INSERT', $sql_ary);
```

Comment:
Do we need to first check if the file is already tracked or do we rely on this being done outside this function?

Reply:
Probably not, this is only executed when the file is not tracked.
Also, if it is, the insert should throw an error because of the unique index i added in the migration

Reply:
This is covered by the unique index key and the exists check in the write method in storage class